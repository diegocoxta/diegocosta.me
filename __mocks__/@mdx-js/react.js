module.exports = {
  MDXProvider: jest.fn(({ children }) => children),
};
