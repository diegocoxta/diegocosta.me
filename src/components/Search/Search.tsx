import React from 'react';
import styled from 'styled-components';
import { Link } from 'gatsby';
import { UseComboboxReturnValue } from 'downshift';
import Fuse from 'fuse.js';

import Container from '~/components/Container';

import { SearchArticlesQuery } from '~/../graphql-types';

const Content = styled.div`
  position: relative;
  margin: 40px 0 20px;
`;

const Input = styled.input`
  transition: box-shadow 0.2s;
  width: 100%;
  padding: 20px;
  background: ${({ theme }) => theme.searchBackgroundColor};
  outline: none;
  color: ${({ theme }) => theme.titleColor};
  font-size: 18px;
  line-height: 18px;
  border-radius: 4px;
  box-sizing: border-box;
  border: 1px solid ${({ theme }) => theme.searchResultsBackgroundColor};

  :focus {
    background-color: ${({ theme }) => theme.searchResultsBackgroundColor};
    box-shadow: 0 10px 20px rgba(0, 0, 0, 0.1);
    color: ${({ theme }) => theme.titleColor};
  }

  ::placeholder {
    color: ${({ theme }) => theme.subtitleColor};
  }

  :focus::placeholder {
    color: ${({ theme }) => theme.subtitleColor};
  }
`;

const Label = styled.label`
  visibility: hidden;
`;

const Results = styled.div`
  position: absolute;
  left: 0;
  right: 0;
  top: calc(20px + 21px + 18px);
  align-items: center;
  cursor: text;
  background: ${({ theme }) => theme.searchResultsBackgroundColor};
  color: black;
  z-index: 4;
  border-bottom-left-radius: 10px;
  border-bottom-right-radius: 10px;
  box-shadow: 0 3px 8px 0 rgba(0, 0, 0, 0.1);
`;

const ResultLink = styled(Link)<{ highlighted: boolean }>`
  display: block;
  text-decoration: none;
  padding: 20px;

  :hover {
    text-decoration: none;
    background: rgba(255, 255, 255, 0.1);
  }
`;

const ResultItemTitle = styled.h3`
  color: ${({ theme }) => theme.titleColor};
  font-size: 16px;
  margin: 0 0 5px;
  font-weight: 400;
`;

const ResultItemDescription = styled.p`
  color: ${({ theme }) => theme.subtitleColor};
  font-size: 14px;
  margin: 0%;
`;

export type ArticleProps = SearchArticlesQuery['articles']['nodes'][0];

interface SearchProps {
  combobox: UseComboboxReturnValue<Fuse.FuseResult<ArticleProps>>;
  articles: Fuse.FuseResult<ArticleProps>[];
}

export default function Search({ articles, combobox }: SearchProps): React.ReactElement {
  return (
    <Container>
      <Content {...combobox.getComboboxProps()}>
        <Input type="text" placeholder="Busque por publicações..." autoComplete="off" {...combobox.getInputProps()} />
        <Label {...combobox.getLabelProps()}>Busque por publicações...</Label>
        <Results {...combobox.getMenuProps()} data-testid="search-results">
          {combobox.isOpen &&
            articles &&
            articles.map((article, index: number) => (
              <ResultLink
                key={article.item.id}
                data-testid="search-results-link"
                to={`/${article?.item?.frontmatter?.language}${article?.item?.fields?.slug}`}
                {...combobox.getItemProps({ index, item: article })}
              >
                <ResultItemTitle data-testid="search-results-title">
                  {article.item.frontmatter?.title ?? ''}
                </ResultItemTitle>
                <ResultItemDescription data-testid="search-results-description">
                  {article.item.frontmatter?.description ?? ''}
                </ResultItemDescription>
              </ResultLink>
            ))}
        </Results>
      </Content>
    </Container>
  );
}
