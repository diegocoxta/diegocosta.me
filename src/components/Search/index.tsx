import React, { useState } from 'react';
import { useStaticQuery, graphql, navigate } from 'gatsby';
import { useCombobox } from 'downshift';
import Fuse from 'fuse.js';

import Search, { ArticleProps } from './Search';

export default (): React.ReactElement => {
  const data = useStaticQuery(graphql`
    query SearchComponent {
      articles: allMarkdownRemark {
        nodes {
          id
          fields {
            slug
          }
          frontmatter {
            title
            description
          }
        }
      }
    }
  `);

  const items = data.articles.nodes;

  const fuse = React.useMemo(
    () =>
      new Fuse(items, {
        minMatchCharLength: 3,
        keys: ['frontmatter.title', 'frontmatter.description'],
      }),
    [items]
  );

  const [articles, setArticles] = useState<Fuse.FuseResult<ArticleProps>[]>([]);

  const combobox = useCombobox({
    items: articles,
    onInputValueChange: ({ inputValue }) => inputValue && setArticles(fuse.search<ArticleProps>(inputValue)),
    itemToString: (article) => article?.item?.frontmatter?.title ?? '',
    onSelectedItemChange: ({ selectedItem }) => navigate(selectedItem?.item?.fields?.slug ?? ''),
  });

  return <Search combobox={combobox} articles={articles} />;
};
