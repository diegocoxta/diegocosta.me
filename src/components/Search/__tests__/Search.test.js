import React from 'react';
import { render } from '@testing-library/react';

import Search from '../Search';

const _MOCKED_COMBOBOX = {
  getComboboxProps: jest.fn(),
  getInputProps: jest.fn(),
  getItemProps: jest.fn(),
  getMenuProps: jest.fn(),
  getLabelProps: jest.fn(),
  isOpen: true,
};

const _MOCKED_ARTICLES = [
  {
    item: {
      id: '90207ca8-97a9-5061-b3a3-8d9b5e5bea7d',
      fields: {
        slug: '/como-migrar-para-o-windows-8-1/',
      },
      frontmatter: {
        title: 'Como migrar para o Windows 8.1 a partir de uma nova instalação do Windows 8',
        description:
          'Quem já precisou formatar um PC e acabou caindo no Windows 8 sabe como é uma dor de cabeça realizar o upgrade para o Windows 8.1.',
      },
    },
    refIndex: 3,
  },
  {
    item: {
      id: '5b82992e-523f-56f7-a0ef-9633a214fd66',
      fields: {
        slug: '/enviando-aplicacao-heroku-windows/',
      },
      frontmatter: {
        title: 'Enviando sua aplicação para o Heroku no Windows',
        description:
          'Quem nunca precisou colocar aquela aplicação em produção para testar ou até mesmo compartilhar com colegas e ficou impedido por não ter um servidor bom para uso gratuito?',
      },
    },
    refIndex: 1,
  },
  {
    item: {
      id: '8d44ee83-477c-527f-b6c4-65028b2ae5c6',
      fields: {
        slug: '/atualizando-programas-cask-homebrew-macos/',
      },
      frontmatter: {
        title: 'Atualizando todos seus casks via Homebrew no macOS',
        description: 'Uma dica simples para forçar a atualização de todos seus casks via HomeBrew',
      },
    },
    refIndex: 4,
  },
];

describe('<Search />', () => {
  it('renders properly', () => {
    const { baseElement, getAllByTestId, getByTestId } = render(
      <Search articles={_MOCKED_ARTICLES} combobox={_MOCKED_COMBOBOX} />
    );

    expect(getByTestId('search-results').firstChild).toBeTruthy();

    expect(getAllByTestId('search-results-link').length).toBe(3);

    _MOCKED_ARTICLES.map((article, index) => {
      expect(getAllByTestId('search-results-title')[index].textContent).toBe(article.item.frontmatter.title);
      expect(getAllByTestId('search-results-description')[index].textContent).toBe(
        article.item.frontmatter.description
      );
      expect(getAllByTestId('search-results-link')[index].href).toContain(article.item.fields.slug);
    });

    expect(_MOCKED_COMBOBOX.getComboboxProps).toHaveBeenCalledTimes(1);
    expect(_MOCKED_COMBOBOX.getLabelProps).toHaveBeenCalledTimes(1);
    expect(_MOCKED_COMBOBOX.getMenuProps).toHaveBeenCalledTimes(1);
    expect(_MOCKED_COMBOBOX.getItemProps).toHaveBeenCalledTimes(3);

    expect(_MOCKED_COMBOBOX.getItemProps).toHaveBeenNthCalledWith(1, { index: 0, item: _MOCKED_ARTICLES[0] });
    expect(_MOCKED_COMBOBOX.getItemProps).toHaveBeenNthCalledWith(2, { index: 1, item: _MOCKED_ARTICLES[1] });
    expect(_MOCKED_COMBOBOX.getItemProps).toHaveBeenNthCalledWith(3, { index: 2, item: _MOCKED_ARTICLES[2] });

    expect(baseElement).toMatchSnapshot();
  });

  it('renders properly with results closed', () => {
    const { baseElement, queryByTestId, getByTestId } = render(
      <Search articles={_MOCKED_ARTICLES} combobox={{ ..._MOCKED_COMBOBOX, isOpen: false }} />
    );

    expect(getByTestId('search-results').firstChild).toBeFalsy();

    expect(queryByTestId('search-results-link')).toBeFalsy();

    expect(baseElement).toMatchSnapshot();
  });
});
