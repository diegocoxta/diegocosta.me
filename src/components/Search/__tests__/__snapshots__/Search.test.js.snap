// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> should render properly 1`] = `
.c0 {
  max-width: 800px;
  margin: 0 auto;
  padding: 0 20px;
}

.c1 {
  position: relative;
  margin: 40px 0 20px;
}

.c2 {
  -webkit-transition: box-shadow 0.2s;
  transition: box-shadow 0.2s;
  width: 100%;
  padding: 20px;
  outline: none;
  font-size: 18px;
  line-height: 18px;
  border-radius: 4px;
  box-sizing: border-box;
  border: 1px solid;
}

.c2 {
  -webkit-transition: box-shadow 0.2s;
  transition: box-shadow 0.2s;
  width: 100%;
  padding: 20px;
  outline: none;
  font-size: 18px;
  line-height: 18px;
  border-radius: 4px;
  box-sizing: border-box;
  border: 1px solid;
}

.c2 {
  -webkit-transition: box-shadow 0.2s;
  transition: box-shadow 0.2s;
  width: 100%;
  padding: 20px;
  outline: none;
  font-size: 18px;
  line-height: 18px;
  border-radius: 4px;
  box-sizing: border-box;
  border: 1px solid;
}

.c2 {
  -webkit-transition: box-shadow 0.2s;
  transition: box-shadow 0.2s;
  width: 100%;
  padding: 20px;
  outline: none;
  font-size: 18px;
  line-height: 18px;
  border-radius: 4px;
  box-sizing: border-box;
  border: 1px solid;
}

.c2:focus {
  box-shadow: 0 10px 20px rgba(0,0,0,0.1);
}

.c3 {
  visibility: hidden;
}

.c4 {
  position: absolute;
  left: 0;
  right: 0;
  top: calc(20px + 21px + 18px);
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: text;
  color: black;
  z-index: 4;
  border-bottom-left-radius: 10px;
  border-bottom-right-radius: 10px;
  box-shadow: 0 3px 8px 0 rgba(0,0,0,0.1);
}

.c5 {
  display: block;
  -webkit-text-decoration: none;
  text-decoration: none;
  padding: 20px;
}

.c5:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  background: rgba(255,255,255,0.1);
}

.c6 {
  font-size: 16px;
  margin: 0 0 5px;
  font-weight: 400;
}

.c7 {
  font-size: 14px;
  margin: 0%;
}

<body>
  <div>
    <section
      class="c0"
    >
      <main>
        <div
          class="c1"
        >
          <input
            autocomplete="off"
            class="c2"
            placeholder="Busque por publicações..."
            type="text"
          />
          <label
            class="c3"
          >
            Busque por publicações...
          </label>
          <div
            class="c4"
            data-testid="search-results"
          >
            <a
              class="c5"
              data-testid="search-results-link"
              href="/undefined/como-migrar-para-o-windows-8-1/"
            >
              <h3
                class="c6"
                data-testid="search-results-title"
              >
                Como migrar para o Windows 8.1 a partir de uma nova instalação do Windows 8
              </h3>
              <p
                class="c7"
                data-testid="search-results-description"
              >
                Quem já precisou formatar um PC e acabou caindo no Windows 8 sabe como é uma dor de cabeça realizar o upgrade para o Windows 8.1.
              </p>
            </a>
            <a
              class="c5"
              data-testid="search-results-link"
              href="/undefined/enviando-aplicacao-heroku-windows/"
            >
              <h3
                class="c6"
                data-testid="search-results-title"
              >
                Enviando sua aplicação para o Heroku no Windows
              </h3>
              <p
                class="c7"
                data-testid="search-results-description"
              >
                Quem nunca precisou colocar aquela aplicação em produção para testar ou até mesmo compartilhar com colegas e ficou impedido por não ter um servidor bom para uso gratuito?
              </p>
            </a>
            <a
              class="c5"
              data-testid="search-results-link"
              href="/undefined/atualizando-programas-cask-homebrew-macos/"
            >
              <h3
                class="c6"
                data-testid="search-results-title"
              >
                Atualizando todos seus casks via Homebrew no macOS
              </h3>
              <p
                class="c7"
                data-testid="search-results-description"
              >
                Uma dica simples para forçar a atualização de todos seus casks via HomeBrew
              </p>
            </a>
          </div>
        </div>
      </main>
    </section>
  </div>
</body>
`;

exports[`<Search /> should render properly with results closed 1`] = `
<body>
  .c0 {
  max-width: 800px;
  margin: 0 auto;
  padding: 0 20px;
}

.c1 {
  position: relative;
  margin: 40px 0 20px;
}

.c2 {
  -webkit-transition: box-shadow 0.2s;
  transition: box-shadow 0.2s;
  width: 100%;
  padding: 20px;
  outline: none;
  font-size: 18px;
  line-height: 18px;
  border-radius: 4px;
  box-sizing: border-box;
  border: 1px solid;
}

.c2 {
  -webkit-transition: box-shadow 0.2s;
  transition: box-shadow 0.2s;
  width: 100%;
  padding: 20px;
  outline: none;
  font-size: 18px;
  line-height: 18px;
  border-radius: 4px;
  box-sizing: border-box;
  border: 1px solid;
}

.c2 {
  -webkit-transition: box-shadow 0.2s;
  transition: box-shadow 0.2s;
  width: 100%;
  padding: 20px;
  outline: none;
  font-size: 18px;
  line-height: 18px;
  border-radius: 4px;
  box-sizing: border-box;
  border: 1px solid;
}

.c2 {
  -webkit-transition: box-shadow 0.2s;
  transition: box-shadow 0.2s;
  width: 100%;
  padding: 20px;
  outline: none;
  font-size: 18px;
  line-height: 18px;
  border-radius: 4px;
  box-sizing: border-box;
  border: 1px solid;
}

.c2:focus {
  box-shadow: 0 10px 20px rgba(0,0,0,0.1);
}

.c3 {
  visibility: hidden;
}

.c4 {
  position: absolute;
  left: 0;
  right: 0;
  top: calc(20px + 21px + 18px);
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: text;
  color: black;
  z-index: 4;
  border-bottom-left-radius: 10px;
  border-bottom-right-radius: 10px;
  box-shadow: 0 3px 8px 0 rgba(0,0,0,0.1);
}

<div>
    <section
      class="c0"
    >
      <main>
        <div
          class="c1"
        >
          <input
            autocomplete="off"
            class="c2"
            placeholder="Busque por publicações..."
            type="text"
          />
          <label
            class="c3"
          >
            Busque por publicações...
          </label>
          <div
            class="c4"
            data-testid="search-results"
          />
        </div>
      </main>
    </section>
  </div>
</body>
`;
